File: src/main/java/org/example/eshop/repository/CartItemRepository.java
package org.example.eshop.repository;

import org.example.eshop.entity.CartItem;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.lang.Nullable;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface CartItemRepository extends JpaRepository<CartItem, Long> {

    List<CartItem> findByCartId(Long cartId);

    List<CartItem> findByVariantId(Long variantId);

    /**
     * May return null when no item exists for the given cartId and variantId.
     */
    @Nullable
    CartItem findByCartIdAndVariantId(Long cartId, Long variantId);

    @Query("SELECT ci FROM CartItem ci WHERE ci.cartId = :cartId")
    List<CartItem> findAllByCartId(@Param("cartId") Long cartId);

    // Derived delete methods (no @Query needed)
    void deleteByCartId(Long cartId);

    void deleteByCartIdAndVariantId(Long cartId, Long variantId);
}

File: src/main/kotlin/org/example/eshop/repository/CartItemRepository.kt
package org.example.eshop.repository

import org.example.eshop.entity.CartItem
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface CartItemRepository : JpaRepository<CartItem, Long> {
    
    fun findByCartId(cartId: Long): List<CartItem>
    
    fun findByVariantId(variantId: Long): List<CartItem>
    
    fun findByCartIdAndVariantId(cartId: Long, variantId: Long): CartItem?
    
    @Query("SELECT ci FROM CartItem ci WHERE ci.cartId = :cartId")
    fun findAllByCartId(@Param("cartId") cartId: Long): List<CartItem>
    
    @Query("DELETE FROM CartItem ci WHERE ci.cartId = :cartId")
    fun deleteByCartId(@Param("cartId") cartId: Long)
    
    @Query("DELETE FROM CartItem ci WHERE ci.cartId = :cartId AND ci.variantId = :variantId")
    fun deleteByCartIdAndVariantId(@Param("cartId") cartId: Long, @Param("variantId") variantId: Long)
}